repository/httpRemoteRepository.go

```diff
@@ -0,0 +1,317 @@
+}
+
+func (r *httpRemote) receiveDoguFromRemoteOrCache(requestUrl string, dirname string) (*core.Dogu, error) {
+	var remoteDogu, err = r.readCachedDogu(dirname)
+	if err != nil {
+		err = r.retrier.Run(func() error {
+			remoteDogu, err = r.requestWithoutCredentialsFirst(requestUrl)
+			if err != nil {
+				remoteDogu, err = r.request(requestUrl, true)
+				return err
+			}
+			return err
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		err = r.writeDoguToCache(remoteDogu, dirname)
+		if err != nil {
+			return &core.Dogu{}, fmt.Errorf("failed to write dogu to cache: %w", err)
+		}
+	}
+
+	return remoteDogu, nil
+}
+
+func (r *httpRemote) readCachedDogu(dirname string) (*core.Dogu, error) {
+	if r.useCache {
+		cacheFile := filepath.Join(dirname, "content.json")
+		doguFromFile, _, err := core.ReadDoguFromFile(cacheFile)
+		if err != nil {
+			return &core.Dogu{}, commonerrors.NewGenericError(fmt.Errorf("failed to read from remote registry and cache: %w", err))
+		}
+		if doguFromFile == nil {
+			return nil, commonerrors.NewNotFoundError(fmt.Errorf("dogu descriptor not found"))
+		}
+		return doguFromFile, nil
+	}
+	return &core.Dogu{}, commonerrors.NewGenericError(fmt.Errorf("useCache is not activated"))
+}
+
+func (r *httpRemote) writeDoguToCache(dogu *core.Dogu, dirname string) error {
+	err := os.MkdirAll(dirname, os.ModePerm)
+	if nil != err {
+		return fmt.Errorf("failed to create cache directory %s: %w", dirname, err)
+	}
+
+	cacheFile := filepath.Join(dirname, "content.json")
+	err = core.WriteDoguToFile(cacheFile, dogu)
+
+	if nil != err {
+		removeErr := os.Remove(cacheFile)
+		if removeErr != nil {
+			core.GetLogger().Warningf("failed to remove cache file %s", cacheFile)
+		}
+		return fmt.Errorf("failed to write cache %s: %w", cacheFile, err)
+	}
+
+	return nil
+}
+
+func (r *httpRemote) requestWithoutCredentialsFirst(requestURL string) (*core.Dogu, error) {
+	core.GetLogger().Debug("Access \"" + requestURL + "\" anonymous...")
+	remoteDogu, err := r.request(requestURL, false)
+	if err != nil {
+		core.GetLogger().Debug("Anonymous access to \"" + requestURL + "\" failed. Using credentials...")
+		remoteDogu, err = r.request(requestURL, true)
+		if err != nil {
+			core.GetLogger().Debug("Access to \"" + requestURL + "\" with credentials failed...")
+		} else {
+			core.GetLogger().Debug("Access to \"" + requestURL + "\" with credentials was successfull...")
+		}
+	}
+	return remoteDogu, err
+}
+
+func (r *httpRemote) request(requestURL string, useCredentials bool) (*core.Dogu, error) {
+	core.GetLogger().Debugf("fetch json from remote %s", requestURL)
+
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return nil, commonerrors.NewGenericError(fmt.Errorf("failed to prepare request: %w", err))
+	}
+
+	if useCredentials && r.credentials != nil {
+		request.SetBasicAuth(r.credentials.Username, r.credentials.Password)
+	}
+
+	resp, err := r.client.Do(request)
```

Comment:
🚫 **[checkstyle]** <bodyclose> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>response body must be closed
<!-- __reviewdog__:ChBlZmVkODU5YjcxZDE1MDBhEgpjaGVja3N0eWxl -->